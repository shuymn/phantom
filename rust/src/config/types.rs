use serde::{Deserialize, Serialize};

/// Main configuration structure for Phantom
#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct PhantomConfig {
    /// Post-create actions to perform after creating a worktree
    #[serde(skip_serializing_if = "Option::is_none")]
    pub post_create: Option<PostCreateConfig>,

    /// Default terminal multiplexer to use
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default_multiplexer: Option<Multiplexer>,
}

/// Post-create configuration
#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct PostCreateConfig {
    /// Files to copy from the original worktree
    #[serde(skip_serializing_if = "Option::is_none")]
    pub copy_files: Option<Vec<String>>,

    /// Commands to run after creating the worktree
    #[serde(skip_serializing_if = "Option::is_none")]
    pub commands: Option<Vec<String>>,
}

/// Supported terminal multiplexers
#[derive(Debug, Clone, Copy, Default, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum Multiplexer {
    Tmux,
    Kitty,
    #[default]
    None,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_serialize_config() {
        let config = PhantomConfig {
            post_create: Some(PostCreateConfig {
                copy_files: Some(vec![".env".to_string(), "config.local.json".to_string()]),
                commands: Some(vec!["npm install".to_string()]),
            }),
            default_multiplexer: Some(Multiplexer::Tmux),
        };

        let json = serde_json::to_string_pretty(&config).unwrap();
        assert!(json.contains("\"postCreate\""));
        assert!(json.contains("\"copyFiles\""));
        assert!(json.contains("\"commands\""));
        assert!(json.contains("\"defaultMultiplexer\""));
        assert!(json.contains("\"tmux\""));
    }

    #[test]
    fn test_deserialize_config() {
        let json = r#"{
            "postCreate": {
                "copyFiles": [".env"],
                "commands": ["npm install"]
            },
            "defaultMultiplexer": "kitty"
        }"#;

        let config: PhantomConfig = serde_json::from_str(json).unwrap();
        assert!(config.post_create.is_some());

        let post_create = config.post_create.unwrap();
        assert_eq!(post_create.copy_files.as_ref().unwrap().len(), 1);
        assert_eq!(post_create.copy_files.unwrap()[0], ".env");
        assert_eq!(post_create.commands.as_ref().unwrap().len(), 1);
        assert_eq!(post_create.commands.unwrap()[0], "npm install");

        assert_eq!(config.default_multiplexer, Some(Multiplexer::Kitty));
    }

    #[test]
    fn test_deserialize_minimal_config() {
        let json = "{}";
        let config: PhantomConfig = serde_json::from_str(json).unwrap();
        assert!(config.post_create.is_none());
        assert!(config.default_multiplexer.is_none());
    }

    #[test]
    fn test_deserialize_multiplexer() {
        assert_eq!(serde_json::from_str::<Multiplexer>("\"tmux\"").unwrap(), Multiplexer::Tmux);
        assert_eq!(serde_json::from_str::<Multiplexer>("\"kitty\"").unwrap(), Multiplexer::Kitty);
        assert_eq!(serde_json::from_str::<Multiplexer>("\"none\"").unwrap(), Multiplexer::None);
    }

    #[test]
    fn test_skip_none_fields() {
        let config = PhantomConfig::default();
        let json = serde_json::to_string(&config).unwrap();
        assert_eq!(json, "{}");
    }
}
