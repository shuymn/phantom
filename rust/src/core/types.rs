use serde::{Deserialize, Serialize};
use std::path::PathBuf;

/// Represents a Git worktree
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Worktree {
    pub name: String,
    pub path: PathBuf,
    pub branch: Option<String>,
    pub commit: String,
    pub is_bare: bool,
    pub is_detached: bool,
    pub is_prunable: bool,
}

/// Git configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GitConfig {
    pub user_name: Option<String>,
    pub user_email: Option<String>,
}

/// Phantom configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PhantomConfig {
    #[serde(default)]
    pub copy_files: Vec<String>,
    #[serde(default)]
    pub terminal: TerminalConfig,
}

/// Terminal configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TerminalConfig {
    #[serde(default = "default_multiplexer")]
    pub multiplexer: String,
}

impl Default for TerminalConfig {
    fn default() -> Self {
        Self { multiplexer: default_multiplexer() }
    }
}

fn default_multiplexer() -> String {
    "auto".to_string()
}

impl Default for PhantomConfig {
    fn default() -> Self {
        Self { copy_files: vec![], terminal: TerminalConfig { multiplexer: default_multiplexer() } }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::Path;

    #[test]
    fn test_worktree_creation() {
        let worktree = Worktree {
            name: "feature-branch".to_string(),
            path: PathBuf::from("/path/to/worktree"),
            branch: Some("feature-branch".to_string()),
            commit: "abc123".to_string(),
            is_bare: false,
            is_detached: false,
            is_prunable: false,
        };

        assert_eq!(worktree.name, "feature-branch");
        assert_eq!(worktree.path, Path::new("/path/to/worktree"));
        assert_eq!(worktree.branch, Some("feature-branch".to_string()));
        assert_eq!(worktree.commit, "abc123");
        assert!(!worktree.is_bare);
        assert!(!worktree.is_detached);
        assert!(!worktree.is_prunable);
    }

    #[test]
    fn test_worktree_serialization() {
        let worktree = Worktree {
            name: "test".to_string(),
            path: PathBuf::from("/test/path"),
            branch: None,
            commit: "def456".to_string(),
            is_bare: true,
            is_detached: true,
            is_prunable: true,
        };

        // Serialize
        let json = serde_json::to_string(&worktree).unwrap();
        assert!(json.contains("\"name\":\"test\""));
        assert!(json.contains("\"is_bare\":true"));

        // Deserialize
        let deserialized: Worktree = serde_json::from_str(&json).unwrap();
        assert_eq!(deserialized.name, worktree.name);
        assert_eq!(deserialized.path, worktree.path);
        assert_eq!(deserialized.is_bare, worktree.is_bare);
    }

    #[test]
    fn test_git_config() {
        let config = GitConfig {
            user_name: Some("John Doe".to_string()),
            user_email: Some("john@example.com".to_string()),
        };

        assert_eq!(config.user_name, Some("John Doe".to_string()));
        assert_eq!(config.user_email, Some("john@example.com".to_string()));

        let empty_config = GitConfig { user_name: None, user_email: None };

        assert!(empty_config.user_name.is_none());
        assert!(empty_config.user_email.is_none());
    }

    #[test]
    fn test_phantom_config_default() {
        let config = PhantomConfig::default();

        assert!(config.copy_files.is_empty());
        assert_eq!(config.terminal.multiplexer, "auto");
    }

    #[test]
    fn test_phantom_config_serialization() {
        let config = PhantomConfig {
            copy_files: vec![".env".to_string(), "config.toml".to_string()],
            terminal: TerminalConfig { multiplexer: "tmux".to_string() },
        };

        // Serialize
        let json = serde_json::to_string_pretty(&config).unwrap();
        assert!(json.contains("\"copy_files\""));
        assert!(json.contains("\".env\""));
        assert!(json.contains("\"multiplexer\": \"tmux\""));

        // Deserialize
        let deserialized: PhantomConfig = serde_json::from_str(&json).unwrap();
        assert_eq!(deserialized.copy_files.len(), 2);
        assert_eq!(deserialized.copy_files[0], ".env");
        assert_eq!(deserialized.terminal.multiplexer, "tmux");
    }

    #[test]
    fn test_terminal_config_default() {
        let config = TerminalConfig::default();
        assert_eq!(config.multiplexer, "auto");
    }

    #[test]
    fn test_terminal_config_custom() {
        let config = TerminalConfig { multiplexer: "kitty".to_string() };
        assert_eq!(config.multiplexer, "kitty");
    }

    #[test]
    fn test_partial_config_deserialization() {
        // Test with minimal JSON
        let json = "{}";
        let config: PhantomConfig = serde_json::from_str(json).unwrap();
        assert!(config.copy_files.is_empty());
        assert_eq!(config.terminal.multiplexer, "auto");

        // Test with only copy_files
        let json = r#"{"copy_files": [".env"]}"#;
        let config: PhantomConfig = serde_json::from_str(json).unwrap();
        assert_eq!(config.copy_files.len(), 1);
        assert_eq!(config.copy_files[0], ".env");
        assert_eq!(config.terminal.multiplexer, "auto");
    }
}
