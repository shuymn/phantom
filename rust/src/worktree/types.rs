use serde::{Deserialize, Serialize};
use std::path::PathBuf;

/// Options for creating a new worktree
#[derive(Debug, Clone, Default)]
pub struct CreateWorktreeOptions {
    /// Branch name to create (defaults to worktree name)
    pub branch: Option<String>,
    /// Commit/branch to base the new worktree on (defaults to HEAD)
    pub commitish: Option<String>,
    /// Files to copy from the source worktree
    pub copy_files: Option<Vec<String>>,
}

/// Result of a successful worktree creation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateWorktreeSuccess {
    pub message: String,
    pub path: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub copied_files: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub skipped_files: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub copy_error: Option<String>,
}

/// Options for deleting a worktree
#[derive(Debug, Clone, Default)]
pub struct DeleteWorktreeOptions {
    /// Force deletion even if there are untracked/modified files
    pub force: bool,
}

/// Result of a successful worktree deletion
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeleteWorktreeSuccess {
    pub message: String,
    pub path: String,
}

/// Result of worktree validation
#[derive(Debug, Clone)]
pub struct WorktreeExistsSuccess {
    pub path: PathBuf,
}

/// Result of worktree non-existence validation
#[derive(Debug, Clone)]
pub struct WorktreeDoesNotExistSuccess {
    pub path: PathBuf,
}
