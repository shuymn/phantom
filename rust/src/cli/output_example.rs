/// Example module showing how insta snapshot testing would be used for CLI output
/// This file demonstrates the pattern to follow when implementing the actual CLI
#[cfg(test)]
mod tests {
    use crate::core::types::Worktree;
    use insta::assert_snapshot;
    use std::path::PathBuf;

    // Example formatter function that would be tested
    fn format_worktree_list(worktrees: &[Worktree]) -> String {
        let mut output = String::new();
        for wt in worktrees {
            output.push_str(&format!(
                "{:<20} {:<50} {}\n",
                wt.name,
                wt.path.display(),
                if wt.is_bare { "(bare)" } else { "" }
            ));
        }
        output
    }

    #[test]
    fn test_format_worktree_list() {
        let worktrees = vec![
            Worktree {
                name: "main".to_string(),
                path: PathBuf::from("/home/user/project"),
                branch: Some("main".to_string()),
                commit: "abc123".to_string(),
                is_bare: false,
                is_detached: false,
                is_locked: false,
                is_prunable: false,
            },
            Worktree {
                name: "feature-xyz".to_string(),
                path: PathBuf::from("/home/user/project/.phantom/feature-xyz"),
                branch: Some("feature/xyz".to_string()),
                commit: "def456".to_string(),
                is_bare: false,
                is_detached: false,
                is_locked: false,
                is_prunable: false,
            },
        ];

        let output = format_worktree_list(&worktrees);
        assert_snapshot!(output);
    }

    #[test]
    fn test_format_empty_list() {
        let worktrees: Vec<Worktree> = vec![];
        let output = format_worktree_list(&worktrees);
        assert_snapshot!(output);
    }

    // Example of testing JSON output format
    fn format_json_output(worktrees: &[Worktree]) -> String {
        serde_json::to_string_pretty(worktrees).unwrap()
    }

    #[test]
    fn test_format_json_output() {
        let worktrees = vec![Worktree {
            name: "test".to_string(),
            path: PathBuf::from("/tmp/test"),
            branch: Some("test-branch".to_string()),
            commit: "123456".to_string(),
            is_bare: false,
            is_detached: false,
            is_locked: false,
            is_prunable: false,
        }];

        let output = format_json_output(&worktrees);
        assert_snapshot!(output);
    }
}

// When implementing the actual CLI, follow this pattern:
// 1. Create formatting functions that take domain objects and return strings
// 2. Use insta::assert_snapshot! to capture the output
// 3. Review the snapshots with `cargo insta review`
// 4. Commit the snapshot files in the `snapshots/` directory
