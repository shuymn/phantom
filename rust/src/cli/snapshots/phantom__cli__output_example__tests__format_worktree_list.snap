---
source: src/cli/output_example.rs
assertion_line: 49
expression: output
---
main                 /home/user/project                                 
feature-xyz          /home/user/project/.phantom/feature-xyz
